Add NSQ shape prediction

PhoenixWorthVCD opened the pull request:
## Description

This is a first pass of partitioning decision which ranks all possible partitions based on the open loop EME data. A reduced set of partitions is forwarded to the closed loop partitioning decision pass.

Closes #721  

## Author

@NaderMahdi 

## Type of Change

New feature

## Tests and Performance

- BD rate difference of -0.05% compared to libaom_fd530f8 in M0 with the fullset of clips.
- No BD rate difference for 360p or screen content clips
- A speed gain of 6% was obtained measured over all clips.
- ~1% memory allocation increase

hassount commented on Source/Lib/Common/Codec/EbDefinitions.h:
```diff
@@ -2891,8 +2922,9 @@ typedef enum EbPictureDepthMode
+    PIC_SB_SWITCH_DEPTH_MODE = 5,  // Adaptive Depth Partitioning
     PIC_OPEN_LOOP_DEPTH_MODE = 4, // Early Inter Depth Decision:  SB size -> 8x8
-    PIC_SB_SWITCH_DEPTH_MODE = 5  // Adaptive Depth Partitioning
+   // PIC_SB_SWITCH_DEPTH_MODE = 5  // Adaptive Depth Partitioning
```
better ordered

PhoenixWorthVCD replied:
Done

hassount commented on Source/Lib/Common/Codec/EbDefinitions.h:
```diff
@@ -145,6 +144,28 @@ enum {
+/* Use open-loop data to predict the NSQ partitions. */
+#define PREDICT_NSQ_SHAPE                               1
+#if PREDICT_NSQ_SHAPE
+#define M3_MDC_LEVEL                                    2
```
is this being used ?

PhoenixWorthVCD replied:
No, it's been removed

hassount commented on Source/Lib/Common/Codec/EbModeDecision.c:
```diff
@@ -719,7 +719,7 @@ EbErrorType mode_decision_candidate_buffer_ctor(
     pictureBufferDescInitData.max_width = MAX_SB_SIZE;
     pictureBufferDescInitData.max_height = MAX_SB_SIZE;
-    pictureBufferDescInitData.bit_depth = max_bitdepth;
+    pictureBufferDescInitData.bit_depth = max_bitdepth; // hardcoded to 8-bit in rtime
```
still a valid comment ?

PhoenixWorthVCD replied:
Still true for rtime branch, but I removed the comment, as it's not hardcoded in master.

hassount commented on Source/Lib/Common/Codec/EbModeDecision.c:
```diff
@@ -4069,60 +4069,63 @@ void inject_new_candidates(
+            MeLcuResults *me_results = picture_control_set_ptr->parent_pcs_ptr->me_results[context_ptr->me_sb_addr];
+
+            EbBool use_close_loop_me = picture_control_set_ptr->parent_pcs_ptr->enable_in_loop_motion_estimation_flag &&
+                ((context_ptr->blk_geom->bwidth == 4 || context_ptr->blk_geom->bheight == 4) || (context_ptr->blk_geom->bwidth > 64 || context_ptr->blk_geom->bheight > 64)) ? EB_TRUE : EB_FALSE;
```
no formatting changes should be combined with the feature commit

hassount commented on Source/Lib/Common/Codec/EbModeDecision.c:
```diff
@@ -4571,14 +4576,14 @@ static TxType intra_mode_to_tx_type(const MbModeInfo *mbmi,
+        PlaneType plane_type,
+        const MacroBlockD *xd, int32_t blk_row,
+        int32_t blk_col, TxSize tx_size,
+        int32_t reduced_tx_set)
```
no formatting

PhoenixWorthVCD replied:
Formatting changes should now be removed.

hassount commented on Source/Lib/Common/Codec/EbModeDecisionConfiguration.c:
```diff
@@ -91,6 +95,12 @@ const uint8_t incrementalCount[85] = {
     0, 0, 0, 4, 0, 0, 0, 4
 };
 
+extern uint32_t get_me_info_index(
```
needs to be flagged

PhoenixWorthVCD replied:
Done

hassount commented on Source/Lib/Common/Codec/EbModeDecisionConfiguration.c:
```diff
@@ -574,6 +583,1143 @@ void MdcInterDepthDecision(
+        tot_cost += split_cost;
+    }
+#endif
+#if 1
```
remove #if 1

PhoenixWorthVCD replied:
Done

hassount commented on Source/Lib/Common/Codec/EbModeDecisionConfiguration.c:
```diff
@@ -574,6 +583,1143 @@ void MdcInterDepthDecision(
+    {
+        tot_cost += context_ptr->local_cu_array[first_blk_idx + blk_it].early_cost;
+    }
+#if 0 //SPLIT_RATE_FIX
```
remove #if 0

PhoenixWorthVCD replied:
done

hassount commented on Source/Lib/Common/Codec/EbModeDecisionConfiguration.c:
```diff
@@ -574,6 +586,1126 @@ void MdcInterDepthDecision(
+    ModeDecisionConfigurationContext         *context_ptr
+)
+{
+    UNUSED(sequence_control_set_ptr);
```
need formatting similar to this uint8_t find_depth_index(uint8_t shape, uint8_t depth_table[NUMBER_OF_DEPTH]) {

PhoenixWorthVCD replied:
Done

hassount commented on Source/Lib/Common/Codec/EbModeDecisionConfiguration.c:
```diff
@@ -574,6 +586,1126 @@ void MdcInterDepthDecision(
+    UNUSED(asm_type);
+    UNUSED(bit_increment);
+    UNUSED(txb_skip_context);
+    UNUSED(dc_sign_context);
```
passing too many arguments to a function then voiding them, can we avoid this ?

hassount commented on Source/Lib/Common/Codec/EbModeDecisionConfiguration.c:
```diff
@@ -574,6 +586,1126 @@ void MdcInterDepthDecision(
+    UNUSED(cr_tu_coeff_bits);
+    UNUSED(txsize_uv);
+    UNUSED(tx_type_uv);
+    UNUSED(asm_type);
```
passing too many arguments to a function then voiding them, can we avoid this ?

hassount commented on Source/Lib/Common/Codec/EbModeDecisionConfiguration.c:
```diff
@@ -574,6 +586,1126 @@ void MdcInterDepthDecision(
+        txb_1d_offset += context_ptr->blk_geom->tx_width[tx_depth][txb_itr] * context_ptr->blk_geom->tx_height[tx_depth][txb_itr];
+    }
+}
+extern void av1_set_ref_frame(MvReferenceFrame *rf,
```
do we need to extern this ? it might create a conflict with aomlib ?

hassount commented on Source/Lib/Common/Codec/EbModeDecisionConfigurationProcess.c:
```diff
@@ -273,6 +273,18 @@ static const int16_t ac_qlookup_12_Q3[QINDEX_RANGE] = {
 };
 
+#if PREDICT_NSQ_SHAPE
+EB_EXTERN EbErrorType nsq_prediction_shape(
```
do we need an extern ?

hassount commented on Source/Lib/Common/Codec/EbPictureDecisionProcess.c:
```diff
@@ -819,27 +819,67 @@ EbErrorType signal_derivation_multi_processes_oq(
-        else
-
-        if (picture_control_set_ptr->enc_mode <= ENC_M2)
+#if PREDICT_NSQ_SHAPE
```
these changes should not be done now unless they're meant to, this feature should only be impacting M0 for now

PhoenixWorthVCD replied:
Done

hassount commented on Source/Lib/Common/Codec/EbPictureDecisionProcess.c:
```diff
@@ -819,27 +819,67 @@ EbErrorType signal_derivation_multi_processes_oq(
+        picture_control_set_ptr->mdc_depth_level = (sequence_control_set_ptr->input_resolution == INPUT_SIZE_576p_RANGE_OR_LOWER) ? MAX_MDC_LEVEL : 6;
+
+    else if (picture_control_set_ptr->enc_mode <= ENC_M2)
+        picture_control_set_ptr->mdc_depth_level = 5;
```
only M0 for now, so this should be removed

PhoenixWorthVCD replied:
done

hassount commented on Source/Lib/Common/Codec/EbPictureDecisionProcess.c:
```diff
@@ -872,14 +912,22 @@ EbErrorType signal_derivation_multi_processes_oq(
+                picture_control_set_ptr->nsq_search_level = NSQ_SEARCH_LEVEL2;
+#else
+				picture_control_set_ptr->nsq_search_level = NSQ_SEARCH_LEVEL3;
+#endif
```
not expected, should be only for M0

PhoenixWorthVCD replied:
Done

hassount commented on Source/Lib/Common/Codec/EbPictureDecisionProcess.c:
```diff
@@ -920,6 +973,20 @@ EbErrorType signal_derivation_multi_processes_oq(
     if (picture_control_set_ptr->pic_depth_mode > PIC_SQ_DEPTH_MODE)
         assert(picture_control_set_ptr->nsq_search_level == NSQ_SEARCH_OFF);
+
+#if NSQ_SUB_LEVEL
```
if this code is not working why do we need it ?

PhoenixWorthVCD replied:
Done - removed

hassount commented on Source/Lib/Common/Codec/EbPictureOperators.c:
```diff
@@ -69,6 +69,74 @@ void pic_copy_kernel_8bit(
         memcpy(dst + j * dst_stride, src + j * src_stride, area_width);
 }
+#if PREDICT_NSQ_SHAPE
+void pic_copy_kernel(
```
seems like duplicate functions, why can't we use 
pic_copy_kernel_8bit
pic_copy_kernel_16bit ?

if they are not, can we link name them differently or add a comment ?

hassount commented on Source/Lib/Common/Codec/EbProductCodingLoop.c:
```diff
@@ -6229,17 +6229,95 @@ void check_redundant_block(const BlockGeom * blk_geom, ModeDecisionContext *cont
+                ret = 0;
+#if NSQ_SUB_LEVEL
+                if (nsq_search_sub_level == NSQ_SEARCH_SUB_LEVEL1) {
+                    if (temporal_layer > 3) {
```
> 3 should not be used, ref / nref

hassount commented on Source/Lib/Common/Codec/EbProductCodingLoop.c:
```diff
@@ -11526,3 +12017,67 @@ EB_EXTERN EbErrorType in_loop_motion_estimation_sblock(
 }
+
+#if PREDICT_NSQ_SHAPE
+uint64_t spatial_full_distortion_helper(
```
this seems to be an sse function when the function does not state so, is there a C equivalent for this ?

hassount commented on Source/Lib/Common/Codec/EbRateDistortionCost.c:
```diff
@@ -1368,6 +1372,290 @@ int svt_is_interintra_allowed(
+    PictureControlSet       *picture_control_set_ptr,
+    CandidateMv             *ref_mv_stack,
+    const BlockGeom         *blk_geom)
+
```
formatting

PhoenixWorthVCD replied:
done

hassount commented on Source/Lib/Common/Codec/EbRateDistortionCost.c:
```diff
@@ -1368,6 +1372,290 @@ int svt_is_interintra_allowed(
+    int16_t           mvRefX;
+    int16_t           mvRefY;
+
+    EbReflist       refListIdx;
```
snake_case

PhoenixWorthVCD replied:
done

hassount commented on Source/Lib/Common/Codec/EbRateDistortionCost.c:
```diff
@@ -1368,6 +1372,290 @@ int svt_is_interintra_allowed(
+
+    uint64_t interModeBitsNum = 0;
+
+    uint8_t skipModeCtx = 0;// cu_ptr->skip_flag_context;
```
snake_case

PhoenixWorthVCD replied:
done

hassount commented on Source/Lib/Common/Codec/EbRateDistortionCost.c:
```diff
@@ -1368,6 +1372,290 @@ int svt_is_interintra_allowed(
+        }
+    }
+
+    uint32_t isInterRate = candidate_ptr->md_rate_estimation_ptr->intra_inter_fac_bits[cu_ptr->is_inter_ctx][1];
```
snake_case

PhoenixWorthVCD replied:
done

hassount commented on Source/Lib/Common/Codec/EbRateDistortionCost.h:
```diff
@@ -192,6 +192,19 @@ extern "C" {
         uint32_t                 top_neighbor_mode);
 
+#if ADD_MDC_FULL_COST
+    extern uint64_t mdc_av1_inter_fast_cost(
```
do we need to extern ?

PhoenixWorthVCD replied:
no, it was removed.

hassount commented on Source/Lib/Common/Codec/EbRateDistortionCost.h:
```diff
@@ -220,6 +233,16 @@ extern "C" {
         BlockSize                                bsize);
 
+#if ADD_MDC_FULL_COST
+    extern int32_t av1_cost_skip_txb(
```
do we need to extern ?

PhoenixWorthVCD replied:
no, it was removed.

hassount commented on Source/Lib/Common/Codec/EbModeDecisionConfigurationProcess.c:
```diff
@@ -2106,7 +2666,40 @@ void* mode_decision_configuration_kernel(void *input_ptr)
+            if (picture_control_set_ptr->parent_pcs_ptr->mdc_depth_level < MAX_MDC_LEVEL) {
+                // SB Constants
+                uint8_t sb_sz = (uint8_t)sequence_control_set_ptr->sb_size_pix;
+                uint8_t lcuSizeLog2 = (uint8_t)Log2f(sb_sz);
```
snake_case

PhoenixWorthVCD replied:
done

PhoenixWorthVCD commented on Source/Lib/Common/Codec/EbModeDecisionConfiguration.c:
```diff
@@ -574,6 +586,1046 @@ void MdcInterDepthDecision(
+            context_ptr->mdc_cu_ptr->compoud_reference_type_context = 0;
+            av1_zero(context_ptr->mdc_cu_ptr->av1xd->neighbors_ref_counts);
+            uint16_t txb_count = context_ptr->blk_geom->txb_count[0];
+            for (uint16_t txb_itr = 0; txb_itr < txb_count; txb_itr++) {
```
done

PhoenixWorthVCD commented on Source/Lib/Common/Codec/EbModeDecisionConfiguration.c:
```diff
@@ -574,6 +586,1046 @@ void MdcInterDepthDecision(
+                results_ptr[parent_depth_idx_mds].early_split_flag = context_ptr->local_cu_array[parent_depth_idx_mds].early_split_flag;
+                last_cu_index = parent_depth_idx_mds;
+            }
+            else {
```
done

PhoenixWorthVCD commented on Source/Lib/Common/Codec/EbModeDecisionConfiguration.c:
```diff
@@ -574,6 +586,1046 @@ void MdcInterDepthDecision(
+    uint32_t first_blk_idx = context_ptr->mds_idx - (context_ptr->blk_geom->totns - 1);//index of first block in this partition
+    uint32_t blk_it;
+
+    for (blk_it = 0; blk_it < context_ptr->blk_geom->totns; blk_it++)
```
done

PhoenixWorthVCD commented on Source/Lib/Common/Codec/EbModeDecisionConfiguration.c:
```diff
@@ -574,6 +586,1046 @@ void MdcInterDepthDecision(
+    return full_cost;
+}
+#endif
+EB_EXTERN EbErrorType nsq_prediction_shape(
```
EB_EXTERN is defined in EbDefinitions.h to be extern "C" if C++ is being used and blank otherwise.

andrey-norkin commented on Source/Lib/Common/Codec/EbDefinitions.h:
```diff
@@ -156,6 +155,27 @@ enum {
+#else
+#define NSQ_TAB_SIZE                                    6
+#endif
+
```
Please remove the extra empty lines

PhoenixWorthVCD replied:
Done

andrey-norkin commented on Source/Lib/Common/Codec/EbFullLoop.h:
```diff
@@ -65,6 +65,21 @@ extern "C" {
         uint64_t                     *y_full_distortion);
 
+#if ADD_MDC_FULL_COST
+    extern void av1_quantize_b_facade_II(
```
A better name for this function (or a comment) would help

PhoenixWorthVCD replied:
This function is already part of the code, and we are just calling it in this PR's changes. Better to change the name in a separate PR.

andrey-norkin commented on Source/Lib/Common/Codec/EbModeDecisionConfiguration.c:
```diff
@@ -574,6 +599,987 @@ void MdcInterDepthDecision(
+extern void av1_quantize_b_facade_II(
+    const TranLow *coeff_ptr,
+    int32_t stride,
+    int32_t                width,
```
a nit: rather long series of spaces between names of these two parameters

PhoenixWorthVCD replied:
Fixed spacing.

andrey-norkin commented on Source/Lib/Common/Codec/EbModeDecisionConfiguration.c:
```diff
@@ -574,6 +599,987 @@ void MdcInterDepthDecision(
+    SequenceControlSet                *sequence_control_set_ptr = (SequenceControlSet*)picture_control_set_ptr->sequence_control_set_wrapper_ptr->object_ptr;
+    EbAsm                             asm_type = sequence_control_set_ptr->encode_context_ptr->asm_type;
+    uint64_t                          y_tu_coeff_bits;
+    uint64_t                          tuFullDistortion[3][DIST_CALC_TOTAL];
```
Different style of the variable name

PhoenixWorthVCD replied:
done

andrey-norkin commented on Source/Lib/Common/Codec/EbModeDecisionConfiguration.c:
```diff
@@ -574,6 +599,987 @@ void MdcInterDepthDecision(
+    SequenceControlSet                *sequence_control_set_ptr,
+    PictureControlSet                 *picture_control_set_ptr,
+    ModeDecisionConfigurationContext  *context_ptr,
+    MdcLcuData                        *mdcResultTbPtr,
```
variable names should be snake_case

PhoenixWorthVCD replied:
Done

andrey-norkin commented on Source/Lib/Common/Codec/EbModeDecisionConfigurationProcess.c:
```diff
@@ -274,6 +274,17 @@ static const int16_t ac_qlookup_12_Q3[QINDEX_RANGE] = {
+    SequenceControlSet                *sequence_control_set_ptr,
+    PictureControlSet                 *picture_control_set_ptr,
+    ModeDecisionConfigurationContext  *context_ptr,
+    MdcLcuData                        *mdcResultTbPtr,
```
variable name should be snake_case

PhoenixWorthVCD replied:
done

andrey-norkin commented on Source/Lib/Common/Codec/EbModeDecisionConfigurationProcess.c:
```diff
@@ -1006,6 +1050,514 @@ void sb_forward_sq_blocks_to_md(
+    uint32_t depth_refinement_mode = AllD;
+    switch (picture_control_set_ptr->parent_pcs_ptr->mdc_depth_level) {
+    case 0:
+        depth_refinement_mode = is_complete_sb ? Pred : AllD;
```
It is not necessarily related to the PR, but these defines: Pred* and AllD should be an enum rather than defines.

PhoenixWorthVCD replied:
Many of those macros were already defined, and are therefore not part of this PR.  Therefore, the new Pred* modes have been kept as macros to be consistent with the existing code base.

andrey-norkin commented on Source/Lib/Common/Codec/EbProductCodingLoop.c:
```diff
@@ -6564,6 +6592,243 @@ PART get_partition_shape(
+    uint8_t ol_part8 = context_ptr->best_nsq_sahpe8;
+    EbBool isCompoundEnabled = (picture_control_set_ptr->parent_pcs_ptr->reference_mode == SINGLE_REFERENCE) ? 0 : 1;
+    uint32_t me_sb_addr;
+    uint32_t me2Nx2NTableOffset;
```
A variable name should be in snake_case

PhoenixWorthVCD replied:
Done

joelsole commented on Source/Lib/Common/Codec/EbModeDecisionConfiguration.c:
```diff
@@ -574,6 +599,987 @@ void MdcInterDepthDecision(
+                for (j = 0; j < context_ptr->blk_geom->tx_height[tx_depth][txb_itr]; j++)
+                    memcpy(rec_buffer + j * candidate_buffer->recon_ptr->stride_y, pred_buffer + j * candidate_buffer->prediction_ptr->stride_y, context_ptr->blk_geom->tx_width[tx_depth][txb_itr]);
+
+                av1_inv_transform_recon8bit(
```
could you add the 10-bit path here?

joelsole replied:
what's the performance for 10-bit?

PhoenixWorthVCD replied:
10-bit is not currently supported for the open loop partitioning decision. It may be supported and evaluated in the future.

hassount replied:
@PhoenixWorthVCD it might be best to turn off MDC for the hbd md for now until this is tested properly

PhoenixWorthVCD replied:
OK, done.  MDC is now off for the hbd md.

hassount replied:
added to #622

hassount commented:
@joelsole @andrey-norkin 
anymore issues with this PR ?

andrey-norkin commented:
In general, you should strongly avoid using a force-push when addressing the review comments. When you do this, the reviewer needs to start looking at the whole PR again rather than going directly to the code around the comments made previously. And this makes the turnaround for your PR slower.

Instead, you should add another commit addressing the comments. You can squash that commit later when your PR is approved if needed.

PhoenixWorthVCD commented:
> In general, you should strongly avoid using a force-push when addressing the review comments. When you do this, the reviewer needs to start looking at the whole PR again rather than going directly to the code around the comments made previously. And this makes the turnaround for your PR slower.
> 
> Instead, you should add another commit addressing the comments. You can squash that commit later when your PR is approved if needed.

Sounds good; I'll avoid doing that next time.
